src/middlewared/middlewared/role.py

```diff
@@ -0,0 +1,62 @@
+from collections import defaultdict
+from dataclasses import dataclass, field
+
+
+@dataclass()
+class Role:
+    includes: [str] = field(default_factory=list)
+
+
+ROLES = {
+    'KEYCHAIN_CREDENTIAL_READ': Role(),
+    'KEYCHAIN_CREDENTIAL_WRITE': Role(includes=['KEYCHAIN_CREDENTIAL_READ']),
+    'REPLICATION_TASK_CONFIG_READ': Role(),
+    'REPLICATION_TASK_CONFIG_WRITE': Role(includes=['REPLICATION_TASK_CONFIG_READ']),
+    'REPLICATION_TASK_READ': Role(),
+    'REPLICATION_TASK_WRITE': Role(includes=['REPLICATION_TASK_READ']),
+    'REPLICATION_TASK_WRITE_PULL': Role(includes=['REPLICATION_TASK_WRITE']),
+    'SNAPSHOT_TASK_READ': Role(),
+    'SNAPSHOT_TASK_WRITE': Role(includes=['SNAPSHOT_TASK_READ']),
+
+    'SNAPSHOT_READ': Role(),
+    'SNAPSHOT_WRITE': Role(includes=['SNAPSHOT_READ']),
+    'SNAPSHOT_DELETE': Role(),
+
+    'REPLICATION_MANAGER': Role(includes=['KEYCHAIN_CREDENTIAL_WRITE',
+                                          'REPLICATION_TASK_CONFIG_WRITE',
+                                          'REPLICATION_TASK_WRITE',
+                                          'SNAPSHOT_TASK_WRITE',
+                                          'SNAPSHOT_WRITE']),
+}
+
+
+class RoleManager:
+    def __init__(self, roles):
+        self.roles = roles
+        self.methods = {}
+        self.allowlists_for_roles = defaultdict(list)
+
+    def register_method(self, method_name, roles):
+        if method_name in self.methods:
+            raise ValueError(f"Method {method_name!r} is already registered in this role manager")
+
+        for role in roles:
+            if role not in self.roles:
+                raise ValueError(f"Invalid role {role!r}")
+
+        self.methods[method_name] = roles
+
+        for role in roles:
+            self.allowlists_for_roles[role].append({"method": "CALL", "resource": method_name})
+
+    def roles_for_fole(self, role):
```

Comment:
`fole` has negative connotation depending on the person's native tongue. I'm  not sure if it has another meaning but nonetheless, it's not a valid method name.